dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java

```diff
@@ -283,7 +295,111 @@ public Token addToken(@ApiParam("code") String satellizerJson) {
 
         final String code = satellizerObject.get("code").getAsString();
 
-        return addGithubToken(code);
+        return addGithubToken(null, code);
+    }
+
+    /**
+     * Adds a Google token to the existing user if user is authenticated already.
+     * Otherwise, create a new Dockstore account too and also add token
+     * @param authUser      The optional Dockstore-authenticated user
+     * @param satellizerJson    Satellizer object returned by satellizer
+     * @return              The user's Dockstore token
+     */
+    @POST
+    @Timed
+    @UnitOfWork
+    @Path("/google")
+    @ApiOperation(value = "Allow satellizer to post a new Google token to dockstore", authorizations = { @Authorization(value = JWT_SECURITY_DEFINITION_NAME) },
+            notes = "A post method is required by satellizer to send the Google token",
+            response = Token.class)
+    public Token addGoogleToken(@ApiParam(hidden = true) @Auth Optional<User> authUser, @ApiParam("code") String satellizerJson) {
+        Gson gson = new Gson();
+        JsonElement element = gson.fromJson(satellizerJson, JsonElement.class);
+        JsonObject satellizerObject = element.getAsJsonObject();
+
+        final String code = satellizerObject.get("code").getAsString();
+        final String redirectUri = satellizerObject.get("redirectUri").getAsString();
+        String accessToken = null;
+        String refreshToken = null;
+        for (int i = 0; i < googleClientID.size() && accessToken == null; i++) {
+            final AuthorizationCodeFlow flow = new AuthorizationCodeFlow.Builder(BearerToken.authorizationHeaderAccessMethod(), HTTP_TRANSPORT,
+                    JSON_FACTORY, new GenericUrl("https://www.googleapis.com/oauth2/v4/token"),
+                    new ClientParametersAuthentication(googleClientID.get(i), googleClientSecret.get(i)), googleClientID.get(i),
+                    "https://accounts.google.com/o/oauth2/v2/auth").build();
+            try {
+                TokenResponse tokenResponse = flow.newTokenRequest(code).setRedirectUri(redirectUri).setRequestInitializer(request -> request.getHeaders().setAccept("application/json")).execute();
+                accessToken = tokenResponse.getAccessToken();
+                refreshToken = tokenResponse.getRefreshToken();
+                LOG.error(tokenResponse.getExpiresInSeconds().toString());
```

Comment:
I'm wondering if we should store the expiration as well? That way when we make calls using the token we can easily know if the access token is expired or not.

Reply:
I don't feel strongly about this. 
For Bitbucket, we just try using the access token and it it doesn't work, we use the refresh token to get a new access token. On one hand, if we're changing one, we should probably change both. On the other hand, storing the expiration is another thing that can go wrong (i.e. if we store it wrong, forget to update it, etc.).